### PR TITLE
[MCCAS] Implement TAG-based materialization of debug info/abbrev

### PR DESCRIPTION
**Pull request by felipepiovezan**

### CODE REVIEW

**Comment by felipepiovezan** [comment]
Once this is merged, I still want to re-root the DIE-based CAS nodes as children of their respective Section CAS blocks

**Comment by felipepiovezan** [comment]
@swift-ci test

File: llvm/include/llvm/MC/CAS/MCCASDebugV1.h [hunk 1]
```diff
@@ -50,6 +50,16 @@ inline uint64_t decodeAbbrevIndexAsAbbrevSetIdx(uint64_t Index) {
...
 }
 
+/// Decodes a value written to the CAS as an index into DWARF's abbreviation
+/// table.
+inline uint64_t decodeAbbrevIndexAsDwarfAbbrevIdx(uint64_t Index) {
+  assert(Index != getEndOfDIESiblingsMarker() &&
+         Index != getDIEInAnotherBlockMarker());
+  // Dwarf 5: Section 7.5.3:
+  // the abbreviation code 0 is reserved for null debugging information entries.
+  return Index - 1;
```
Comment (felipepiovezan) [inline llvm/include/llvm/MC/CAS/MCCASDebugV1.h:60]: This -1 may look mysterious here, but please look at it in the context of the functions right above it.

**Comment by felipepiovezan** [comment]
Rebased, changed some of the tests that still expected the old impl

**Comment by felipepiovezan** [comment]
@swift-ci test

File: llvm/lib/MC/MCCASDebugV1.cpp [hunk 2]
```diff
@@ -264,3 +264,33 @@ Expected<dwarf::Form> AbbrevEntryReader::readForm() {
...
   return static_cast<dwarf::Form>(FormAsInt);
 }
+
+uint64_t
+mccasformats::v1::reconstructAbbrevSection(raw_ostream &OS,
+                                           ArrayRef<StringRef> AbbrevEntries) {
+  uint64_t WrittenSize = 0;
+  for (auto [EntryIdx, EntryData] : enumerate(AbbrevEntries)) {
+    // Dwarf 5: Section 7.5.3:
+    // Each declaration begins with an unsigned LEB128 number representing the
```
Comment (rastogishubham) [inline llvm/lib/MC/MCCASDebugV1.cpp:276]: Nice comment!

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 3]
```diff
@@ -690,6 +690,29 @@ materializeDebugInfoSection(MCCASReader &Reader, ArrayRef<cas::ObjectRef> Refs,
                                   LoadedString->MapOfStringOffsets, &Reader.OS);
 }
 
+static Expected<DIETopLevelRef>
+findDIETopLevelRef(MCCASReader &Reader, ArrayRef<cas::ObjectRef> Refs) {
```
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp:486]: If I remember correctly, the `DIETopLevelRef` sits just below the `SectionRef` right? Do we need to loop over `Refs` to find it? Wouldn't there just be one `Ref` under `DebugAbbrevSectionRef` that is a `DIETopLevelRef`? And so on for all the sections we want to materialize?

File: llvm/include/llvm/MC/CAS/MCCASDebugV1.h (hunk 1 above)
Comment (rastogishubham) [reply llvm/include/llvm/MC/CAS/MCCASDebugV1.h:60]: It makes sense

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 4]
```diff
@@ -723,6 +723,46 @@ Expected<uint64_t> materializeAbbrevFromTagImpl(MCCASReader &Reader,
   return reconstructAbbrevSection(Reader.OS, LoadedTopRef->AbbrevEntries);
 }
 
+static Expected<uint64_t>
+materializeDebugInfoFromTagImpl(MCCASReader &Reader,
+                                ArrayRef<cas::ObjectRef> Refs,
+                                DebugInfoSectionRef SectionRef) {
+  Expected<DIETopLevelRef> MaybeTopRef = findDIETopLevelRef(Reader, Refs);
```
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp:512]: Same as my earlier comment, I thought that the `DebugInfoSectionRef` will only have one ref below it which will be the `DIETopLevelRef`.

On another note, it seems like the `DIETopLevelRef` is redundant and can just be removed?

File: llvm/test/DebugInfo/CAS/AArch64/debug_line_and_info_dedupe.test [hunk 5]
```diff
@@ -3,24 +3,23 @@ RUN: split-file %s %t
 
 RUN: llc -O0 --filetype=obj --cas-backend --cas=%t/cas --mccas-casid %t/a.ll -o %t/a.id
 RUN: llc -O0 --filetype=obj --cas-backend --cas=%t/cas --mccas-casid %t/b.ll -o %t/b.id
-RUN: llvm-cas-dump --cas=%t/cas --casid-file %t/a.id %t/b.id | FileCheck %s
+RUN: llvm-cas-dump --die-refs --cas=%t/cas --casid-file %t/a.id %t/b.id | FileCheck %s
```
Comment (rastogishubham) [inline llvm/test/DebugInfo/CAS/AArch64/debug_line_and_info_dedupe.test:6]: Nice!

File: llvm/test/DebugInfo/CAS/AArch64/test_relocation_zero.ll [hunk 6]
```diff
@@ -1,16 +1,14 @@
...
-;CHECK: mc:debug_info_cu llvmcas://{{[0-9a-z]+}}
-;CHECK-NEXT: 0x00 0x00 0x00 0x00
-;CHECK-SAME: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
-;CHECK-SAME: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
-;CHECK: mc:debug_info_cu llvmcas://{{[0-9a-z]+}}
-;CHECK-NEXT: 0x00 0x00 0x00 0x00
-;CHECK-SAME: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
-;CHECK-SAME: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
+; CHECK-NOT: FORM_addr
+; CHECK:  DW_FORM_addr               [dedups]   [0 0 0 0 0 0 0 0]
```
Comment (rastogishubham) [inline llvm/test/DebugInfo/CAS/AArch64/test_relocation_zero.ll:8]: Nice!

**Review by rastogishubham — APPROVED** [review]
It looks good overall! But I do have some comments inline that may need be addressed in a future patch

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 3 above)
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp:486]: Yeah, as I mentioned in the first comment of this PR, there is still some cleanup to be done in future patches:

> Once this is merged, I still want to re-root the DIE-based CAS nodes as children of their respective Section CAS blocks

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 4 above)
Comment (felipepiovezan) [reply llvm/lib/MC/MCCASObjectV1.cpp:512]: Yup, there is some cleanup to be done!

**Comment by felipepiovezan** [comment]
rebased

**Comment by felipepiovezan** [comment]
@swift-ci test

File: llvm/lib/MC/MCCASObjectV1.cpp (hunk 3 above)
Comment (rastogishubham) [reply llvm/lib/MC/MCCASObjectV1.cpp:486]: Sounds good

**Comment by felipepiovezan** [comment]
@swift-ci test

**Pull request merged**